usr/src/lib/smbclnt/libfknsmb/common/fake_stream.c

```diff
@@ -0,0 +1,1834 @@
+{
+	mblk_t *tmp;
+	mblk_t *lastp = NULL;
+
+	ASSERT(mp && bp);
+	for (tmp = mp; tmp; tmp = tmp->b_cont) {
+		if (tmp == bp) {
+			if (lastp)
+				lastp->b_cont = tmp->b_cont;
+			else
+				mp = tmp->b_cont;
+			tmp->b_cont = NULL;
+			return (mp);
+		}
+		lastp = tmp;
+	}
+	return ((mblk_t *)-1);
+}
+
+/*
+ * Concatenate and align first len bytes of common
+ * message type.  Len == -1, means concat everything.
+ * Returns 1 on success, 0 on failure
+ * After the pullup, mp points to the pulled up data.
+ */
+int
+pullupmsg(mblk_t *mp, ssize_t len)
+{
+	mblk_t *bp, *b_cont;
+	dblk_t *dbp;
+	ssize_t n;
+
+	ASSERT(mp->b_datap->db_ref > 0);
+	ASSERT(mp->b_next == NULL && mp->b_prev == NULL);
+
+	/*
+	 * We won't handle Multidata message, since it contains
+	 * metadata which this function has no knowledge of; we
+	 * assert on DEBUG, and return failure otherwise.
+	 */
+	ASSERT(mp->b_datap->db_type != M_MULTIDATA);
+	if (mp->b_datap->db_type == M_MULTIDATA)
+		return (0);
+
+	if (len == -1) {
+		if (mp->b_cont == NULL && str_aligned(mp->b_rptr))
+			return (1);
+		len = xmsgsize(mp);
+	} else {
+		ssize_t first_mblk_len = mp->b_wptr - mp->b_rptr;
+		ASSERT(first_mblk_len >= 0);
+		/*
+		 * If the length is less than that of the first mblk,
+		 * we want to pull up the message into an aligned mblk.
+		 * Though not part of the spec, some callers assume it.
+		 */
+		if (len <= first_mblk_len) {
+			if (str_aligned(mp->b_rptr))
+				return (1);
+			len = first_mblk_len;
+		} else if (xmsgsize(mp) < len)
+			return (0);
+	}
+
+	if ((bp = allocb_tmpl(len, mp)) == NULL)
+		return (0);
+
+	dbp = bp->b_datap;
+	*bp = *mp;		/* swap mblks so bp heads the old msg... */
+	mp->b_datap = dbp;	/* ... and mp heads the new message */
+	mp->b_datap->db_mblk = mp;
+	bp->b_datap->db_mblk = bp;
+	mp->b_rptr = mp->b_wptr = dbp->db_base;
+
+	do {
+		ASSERT(bp->b_datap->db_ref > 0);
+		ASSERT(bp->b_wptr >= bp->b_rptr);
+		n = MIN(bp->b_wptr - bp->b_rptr, len);
+		ASSERT(n >= 0);		/* allow zero-length mblk_t's */
+		if (n > 0)
+			bcopy(bp->b_rptr, mp->b_wptr, (size_t)n);
+		mp->b_wptr += n;
+		bp->b_rptr += n;
+		len -= n;
+		if (bp->b_rptr != bp->b_wptr)
+			break;
+		b_cont = bp->b_cont;
+		freeb(bp);
+		bp = b_cont;
+	} while (len && bp);
+
+	mp->b_cont = bp;	/* tack on whatever wasn't pulled up */
+
+	return (1);
+}
+
+/*
+ * Concatenate and align at least the first len bytes of common message
+ * type.  Len == -1 means concatenate everything.  The original message is
+ * unaltered.  Returns a pointer to a new message on success, otherwise
+ * returns NULL.
+ */
+mblk_t *
+msgpullup(mblk_t *mp, ssize_t len)
+{
+	mblk_t	*newmp;
+	ssize_t	totlen;
+	ssize_t	n;
+
+	/*
+	 * We won't handle Multidata message, since it contains
+	 * metadata which this function has no knowledge of; we
+	 * assert on DEBUG, and return failure otherwise.
+	 */
+	ASSERT(mp->b_datap->db_type != M_MULTIDATA);
+	if (mp->b_datap->db_type == M_MULTIDATA)
+		return (NULL);
```

Comment:
I really really want to kill these tests from the kernel proper.  These stupid if tests run on so many mblk operations...

Reply:
I cut out most of the unused stuff.  This one seemed small enough so I didn't bother.